### PR TITLE
[helpers TS conversion] classes

### PR DESCRIPTION
**Pull request by amjed-98**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes Part of #16500 
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | No
| License                  | MIT

Migrate all **classes** helper functions (createClass, possibleConstructorReturn, classCallCheck, assertThisInitialized) to **TypeScript**

### CODE REVIEW

**Comment by liuxingbaoyu** [comment]
Thanks! For compatibility, we need to keep `var`. :)

File: packages/babel-helpers/src/helpers/assertThisInitialized.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /* @minVersion 7.0.0-beta.0 */
 
-export default function _assertThisInitialized(self) {
+export default function _assertThisInitialized<T>(self: T): T {
```
Comment (SukkaW) [inline packages/babel-helpers/src/helpers/assertThisInitialized.ts:3]:
```suggestion
export default function _assertThisInitialized<T>(self: T | undefined): T {
```

`_assertThisInitialized` does accept `undefined` though.

File: packages/babel-runtime/helpers/possibleConstructorReturn.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+var _typeof = require("./typeof.js")["default"];
+var assertThisInitialized = require("./assertThisInitialized.js");
+
+function _possibleConstructorReturn<T, U>(t: T, e: U): U | T {
```
Comment (SukkaW) [inline packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: Is it possible to migrate to `export default`? Let's give it a shot and see if the `make test` passes.

Comment (amjed-98) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: great!
on it

Comment (amjed-98) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: should I also convert `require()` to use `esm import?`
@SukkaW

Comment (liuxingbaoyu) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: The files in this directory should be automatically generated and we should not edit it manually.

**Comment by amjed-98** [comment]
> For compatibility, we need to keep `var`. :)

I must have changed it out of habit unconsciously 

reverted! @liuxingbaoyu

**Comment by liuxingbaoyu** [comment]
Did you forget to push your changes? :)

**Comment by amjed-98** [comment]
pushed

File: packages/babel-runtime/helpers/possibleConstructorReturn.ts (hunk 2 above)
Comment (amjed-98) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: oh yeah you're right

there is another file called possibleConstructorReturn.js in packages/babel-helpers/src/helpers/possibleConstructorReturn.js

which has the same name as this one

should I also convert it to typescript?

Comment (liuxingbaoyu) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts:4]: Yes, we should convert it and remove this.

Comment (amjed-98) [reply packages/babel-runtime/helpers/possibleConstructorReturn.ts]: fixed and pushed

File: packages/babel-helpers/src/helpers/createClass.ts [hunk 3]
```diff
@@ -1,7 +1,13 @@
 /* @minVersion 7.0.0-beta.0 */
 
-import toPropertyKey from "toPropertyKey";
-function _defineProperties(target, props) {
+// @ts-expect-error Migrate in another PR
+import toPropertyKey from "toPropertyKey.ts";
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/createClass.ts:4]:
```suggestion
import toPropertyKey from "./toPropertyKey.ts";
```

(and it's already written in TS :) )

File: packages/babel-helpers/src/helpers/possibleConstructorReturn.ts [hunk 4]
```diff
@@ -1,8 +1,10 @@
 /* @minVersion 7.0.0-beta.0 */
 
-import assertThisInitialized from "assertThisInitialized";
+import assertThisInitialized from "./assertThisInitialized.ts";
 
-export default function _possibleConstructorReturn(self, call) {
+export default function _possibleConstructorReturn<
+  T extends object | undefined,
+>(self: T, call: T | ((...args: any[]) => T) | undefined) {
```
Comment (nicolo-ribaudo) [inline packages/babel-helpers/src/helpers/possibleConstructorReturn.ts:7]: While we are here, let's rename `call` to `value`. Calling it `call` makes absolutely no sense 😅 

It's the value passed to `return` in a derived class constructor:
```js
class A extends B {
  constructor() {
    return 1;
  }
}
```

"call" is `1` in this case.

(and probably it's type should just be `any` or `unknown`)

Comment (amjed-98) [reply packages/babel-helpers/src/helpers/possibleConstructorReturn.ts:7]: ok!!
that makes much sense now!

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Review by liuxingbaoyu — APPROVED** [review]

**Pull request merged**